lib/components/form/batch-settings.tsx

```diff
@@ -71,57 +75,43 @@ type Props = {
   config: any
   currentQuery: any
   intl: IntlShape
-  modeOptions: Mode[]
-  possibleCombinations: Combination[]
+  modeButtonOptions: ModeButtonDefinition[]
   routingQuery: any
   setQueryParam: (queryParam: any) => void
 }
 
-type State = {
-  expanded?: string | null
-  selectedModes: string[]
-}
-
 /**
  * Main panel for the batch/trip comparison form.
  */
-class BatchSettings extends Component<Props, State> {
-  constructor(props: Props) {
-    super(props)
-    this.state = {
-      expanded: null,
-      selectedModes: getDefaultModes(props.modeOptions) || []
-    }
-  }
-
-  _onClickMode = (mode: string) => {
-    const { currentQuery, modeOptions, possibleCombinations, setQueryParam } =
-      this.props
-    const { selectedModes } = this.state
-    const index = selectedModes.indexOf(mode)
-    const enableMode = index === -1
-    const newModes = [...selectedModes]
-    if (enableMode) newModes.push(mode)
-    else newModes.splice(index, 1)
-    // Update selected modes for mode buttons.
-    this.setState({ selectedModes: newModes })
-    // Update the available mode combinations based on the new modes selection.
-    const possibleModes = modeOptions.map((m) => m.mode)
-    const disabledModes = possibleModes.filter((m) => !newModes.includes(m))
-    // Only include a combination if it every required mode is enabled.
-    const newCombinations = possibleCombinations
-      .filter(combinationFilter(newModes))
-      .map(replaceTransitMode(currentQuery.mode))
-
-    setQueryParam({
-      combinations: newCombinations,
-      disabledModes,
-      enabledModes: newModes
-    })
-  }
+function BatchSettings({
+  activeSearch,
+  config,
+  currentQuery,
+  intl,
+  routingQuery
+}: Props) {
+  const [expanded, setExpanded] = useState<null | string>(null)
+  const { ModeIcon } = useContext(ComponentContext)
+
+  const modeButtonsWithIcons: ModeButtonDefinition[] =
+    config.modes.modeButtons.map((button: ModeButtonDefinition) => ({
+      ...button,
+      Icon: React.memo(function ModeButtonIcon() {
+        return <ModeIcon mode={button.iconName} />
+      })
+    }))
```

Comment:
Can these be moved to `mapStateToProps`? I think redux applies an memo or some optimization that prevents too many frequent changes.

Reply:
I don't think we can do that because of the use of ModeIcon from context, which is only available in a component. I believe I looked into putting this in the `mapStateToProps` when I wrote it.